tests/test_branch.rs

```diff
@@ -0,0 +1,415 @@
+extern crate narm;
+mod common;
+
+use common::*;
+use narm::narmvm::*;
+
+/*
+
+Unit test for Branch operators
+
+Included varieties:
+
+B <label> T2            - Branch unconditionally by label, jump size restricted to signed 11 bits
+B<C> <label> T1         - Branch conditionally by label, jump size restricted to signed 8 bits
+BX <Rm> T1              - Branch by register
+BL <label> T1           - Branch by label, set link register, jump size restricted to even signed 25 bits (?????)
+BLX <Rm> T1             - Branch by register, set link register
+
+General test cases:
+
+- Branch forward
+- Branch backward
+- Branch far forward (Causing memory error)
+- Branch far backward (Causing memory error)
+
+Special test case for BLX <Rm> T1:
+
+- Branch and then branch back using address saved in link register
+
+Special test case for B<C>:
+
+- Test all different conditions
+
+The reference for these tests is currently official documentations and a QEMU-based VM
+TODO: Test against a hardware Cortex-M0 to make sure it's actually up to spec?
+
+*/
+
+// String representation of ops for use in debug output
+const OPCODES: &'static [&'static str] = &[
+    "B <label> T2",
+    "B<C> <label> T1",
+    "BX <Rm> T1",
+    "BL <label> T1 32bit",
+    "BLX <Rm> T1",
+];
+
+// Simple constant for number of opcodes tested in this file
+const NUM_OPCODES: &'static usize = &5;
+
+// Branch forward
+#[test]
+pub fn test_branch_forward() {
+    println!("\n>>> Branch ops test case: Branch forward \n");
+
+    // Arrays holding instances of VMs and matching state structs
+    let mut vms: [NarmVM; *NUM_OPCODES] = Default::default();
+    let mut vm_states: [VMState; *NUM_OPCODES] = Default::default();
+
+    // Tell macros which op varieties are tested in this function
+    let applicable_op_ids = vec![0, 1, 2, 3, 4];
+
+    // Common pre-execution state
+    common_state!(
+        applicable_op_ids,
+        vm_states.r[1] = Some(ASM_ENTRY + 2 * OP_SIZE + THUMBS_MODE)
+    );
+
+    // VM initialization
+    let post_ops = "
+        svc             #0x01
+        test1:
+        movs r0,        #0xCD
+        svc             #0xFF
+        ";
+
+    // 0: B <label> T2"
+    let ops0 = format!("b test1 {}", post_ops);
+    create_vm!(vms, vm_states, 0, code_var = true, ops0);
```

Comment:
What is `code_var = true`? Looking through the macro it's not obvious what the purpose is and the syntax is a bit foreign here

Reply:
Macro syntax is still very much up in the air. That line is really just a token to help both the test writer and the macro to know what macro pattern to use. I've intended to come up with something smarter anyway though, so I'll look into it.